Sources/MongoSwift/ReadWriteConcern.swift

```diff
@@ -0,0 +1,91 @@
+import libmongoc
+
+/// An enumeration of possible ReadConcern levels.
+public enum ReadConcernLevel: String {
+    case local
+    case available
+    case majority
+    case linearizable
+    case snapshot
+}
+
+/// A class to represent a MongoDB read concern.
+public class ReadConcern {
+
+    /// A pointer to a mongoc_read_concern_t
+    internal var _readConcern: OpaquePointer?
+
+    /// The level of this readConcern, or nil if the level is not set.
+    public var level: String? {
+        guard let level = mongoc_read_concern_get_level(self._readConcern) else {
+            return nil
+        }
+        return String(cString: level)
+    }
+
+    /// Initialize a new ReadConcern from a ReadConcernLevel.
+    public convenience init(_ level: ReadConcernLevel) throws {
+        try self.init(level.rawValue)
+    }
+
+    /// Initialize a new ReadConcern from a String corresponding to a read concern level.
+    public init(_ level: String) throws {
+        self._readConcern = mongoc_read_concern_new()
+        if !mongoc_read_concern_set_level(self._readConcern, level) {
+            throw MongoError.readConcernError(message: "Failed to set read concern level to '\(level)'")
+        }
+    }
+
+    /// Initialize a new empty ReadConcern.
+    public init() {
+        self._readConcern = mongoc_read_concern_new()
+    }
+
+    /// Initialize a new ReadConcern from a Document.
+    public convenience init(_ doc: Document) throws {
+        if let level = doc["level"] as? String {
+            try self.init(level)
+        } else {
+            self.init()
+        }
+    }
+
+    /// Initializes a new ReadConcern by copying a mongoc_read_concern_t.
+    /// The caller is responsible for freeing the original mongoc_read_concern_t.
+    internal init(_ readConcern: OpaquePointer?) {
+        self._readConcern = mongoc_read_concern_copy(readConcern)
+    }
+
+    /// Appends this readConcern to a Document.
+    internal func append(to doc: Document) throws {
+        if !mongoc_read_concern_append(self._readConcern, doc.data) {
+            throw MongoError.readConcernError(message: "Error appending readconcern to document \(doc)")
+        }
+    }
+
+    /// Since we have to follow certain rules about whether to include or omit a ReadConcern,
+    /// this function handles obeying those, factoring in the ReadConcern, if any, for
+    /// whatever object is calling this function. It returns a final options Document for the
+    /// calling function to use, or nil if the Document ends up being empty.
+    internal static func append(_ readConcern: ReadConcern?, to opts: Document?, callerRC: ReadConcern?) throws -> Document? {
```

Comment:
I was initially hoping that we could achieve appending the `ReadConcern` by just making `ReadConcern` a `BsonEncodable` and handling the logic in its `encode` function, but given that whether we should encode it depends on the context we're calling it from (i.e. whether the user has set a `ReadConcern` already), I wasn't sure how it would be possible to build that logic in